build_runner_core/lib/src/util/build_dirs.dart

```diff
@@ -4,18 +4,27 @@
 
 import 'package:build/build.dart';
 
+import '../generate/options.dart';
 import '../generate/phase.dart';
 
-/// Returns whether or not [id] should be built based upon [buildDirs] and
-/// [phase].
-bool shouldBuildForDirs(AssetId id, List<String> buildDirs, BuildPhase phase) {
+/// Returns whether or not [id] should be built based upon [buildDirs],
+/// [phase], and optional [buildFilters].
+bool shouldBuildForDirs(AssetId id, List<String> buildDirs, BuildPhase phase,
+    {Iterable<BuildFilter> buildFilters}) {
   // Always build non-hidden outputs, since they ship with the package.
   // Otherwise it is too easy to ship not up to date files on accident.
   if (!phase.hideOutput) return true;
 
   // Always build `lib/` dirs, anything in them might be available to any app.
   if (id.path.startsWith('lib/')) return true;
```

Comment:
Does this imply that the main entrypoints of a flutter application will always be built? (due to existing under lib)

Reply:
It does, yes :sob: .

I don't think we can change this unfortunately. Sass is a really concrete example - although this entire feature could have other issues with sass too, but this at least helps paper over some of the problems.

Note that this is only for `required` builders still (but web entrypoints are one of those). I really don't think we can avoid building required builders from dependencies because we have no concept of a RUNFILES equivalent (ie: what files are required at runtime for a given app).

Reply:
Well, I guess they will get cached at least. Ideally we'd have a way to avoid multiple dart2js invocations for an app with more than one entrypoint, but as long as we can skip compiling extra tests this seems like a win.


Of course, dart2js itself could be more modular...

Reply:
See my recent comment https://github.com/dart-lang/build/pull/2415#issuecomment-522083547.

For flutter web since it doesn't support custom builders anyways it is well known that you don't have to actually build these lib files, and that proposal would allow us to delete that code and have webdev explicitly opt into it instead.